rebuild/src/public_key.cc

```diff
@@ -0,0 +1,184 @@
+    }
+    // Convert to final Uint8Array type and check length is valid
+    Napi::Uint8Array pk_bytes = pk_array.As<Napi::TypedArrayOf<uint8_t>>();
+    std::string err_out{"pkBytes"};
+    if (!is_valid_length(
+            err_out,
+            pk_bytes.ByteLength(),
+            BLST_TS_PUBLIC_KEY_LENGTH_COMPRESSED,
+            BLST_TS_PUBLIC_KEY_LENGTH_UNCOMPRESSED))
+    {
+        Napi::TypeError::New(env, err_out).ThrowAsJavaScriptException();
+        return scope.Escape(env.Undefined());
+    }
+    // Get module for globals and run PublicKey constructor
+    BlstTsAddon *module = env.GetInstanceData<BlstTsAddon>();
+    // Pass void External to constructor so can tell if constructor was called
+    // from C or JS to prevent direct calls from JS to ensure proper setup
+    Napi::Object wrapped = module->_public_key_ctr.New({Napi::External<void>::New(env, nullptr)});
+    // Setup object correctly.  Start with type tagging wrapper class.
+    wrapped.TypeTag(&module->_public_key_tag);
+    // Unwrap object to get native instance
+    PublicKey *pk = PublicKey::Unwrap(wrapped);
+    // default to jacobian
+    pk->_has_jacobian = true;
+
+    // but figure out if request for affine
+    if (!info[1].IsUndefined())
+    {
+        Napi::Value type_val = info[1].As<Napi::Value>();
+        if (!type_val.IsNumber())
+        {
+            Napi::TypeError::New(env, "type must be of enum CoordType (number)").ThrowAsJavaScriptException();
+            return scope.Escape(env.Undefined());
+        }
+        if (type_val.As<Napi::Number>().Uint32Value() == 0)
+        {
+            pk->_has_jacobian = false;
+            pk->_has_affine = true;
+        }
+    }
+
+    try
+    {
+        if (pk->_has_jacobian)
+        {
+            pk->_jacobian.reset(new blst::P1{pk_bytes.Data(), pk_bytes.ByteLength()});
+        }
+        else
+        {
+            pk->_affine.reset(new blst::P1_Affine{pk_bytes.Data(), pk_bytes.ByteLength()});
+        }
+    }
+    catch (blst::BLST_ERROR err)
+    {
+        Napi::RangeError::New(env, module->GetBlstErrorString(err)).ThrowAsJavaScriptException();
+        return scope.Escape(env.Undefined());
+    }
+
+    /**
+     * TODO: Check for zero key for spec tests
+     */
+
+    return scope.Escape(wrapped);
+}
+
+PublicKey::PublicKey(const Napi::CallbackInfo &info)
+    : Napi::ObjectWrap<PublicKey>{info},
+      _has_jacobian{false},
+      _has_affine{false},
+      _jacobian{nullptr},
+      _affine{nullptr}
+{
+    Napi::Env env = info.Env();
+    Napi::HandleScope scope(env);
+    // Check that constructor was called from C++ and not JS. Externals can only
+    // be created natively.
+    if (!info[0].IsExternal())
+    {
+        Napi::Error::New(env, "PublicKey constructor is private").ThrowAsJavaScriptException();
+        return;
+    }
+};
+
+Napi::Value PublicKey::Serialize(const Napi::CallbackInfo &info)
+{
+    Napi::Env env = info.Env();
+    Napi::EscapableHandleScope scope(env);
```

Comment:
for my understanding what does `Escape` do?

Reply:
It allows the handle to "escape" to the surrounding context. Without this the object handle that is returned from the function would be considered part of the current function context from a lexical perspective and would be allowed to be GC'd after the function returned.  When the value is escaped it is telling the GC that the object should live for the lifetime of the context that it is returned to instead. 

In practice it holds the handle reference and appends it to the end of the outer handle scope list once the exiting scope is popped off as the active context.

https://github.com/nodejs/node/blob/4166d40d0873b6d8a0c7291872c8d20dc680b1d7/deps/v8/src/api/api.cc#L948

Reply:
As additional context the "scope" is a list of addresses to heap objects.  As values are created they are added to the end of the list in lexical order and the "start slot" for the current context is maintained by the isolate. When a function returns and the scope is no longer relevant the marker is moved back the the last start/end slot for the outer context and everything that is after the end is considered out of scope. The escapee is appended to the end and the end pointer is incremented to include the escapee so its "still in scope" and not collectable.